2. Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |



In the context of this problem, what is a false positive?

7 false positives where predicted cat, but was actual dog.

In the context of this problem, what is a false negative?

13 false negatives where predicted dog, but actual cat.

How would you describe this model?

3. You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [2]:
import pandas as pd
from sklearn.metrics import confusion_matrix


In [3]:
c3_df = pd.read_csv('c3.csv')

In [4]:
c3_df

actual     model1     model2     model3
0    No Defect  No Defect     Defect  No Defect
1    No Defect  No Defect     Defect     Defect
2    No Defect  No Defect     Defect  No Defect
3    No Defect     Defect     Defect     Defect
4    No Defect  No Defect     Defect  No Defect
..         ...        ...        ...        ...
195  No Defect  No Defect     Defect     Defect
196     Defect     Defect  No Defect  No Defect
197  No Defect  No Defect  No Defect  No Defect
198  No Defect  No Defect     Defect     Defect
199  No Defect  No Defect  No Defect     Defect

[200 rows x 4 columns]

In [5]:
c3_df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [16]:
pd.crosstab(c3_df.model1, c3_df.actual)

actual     Defect  No Defect
model1                      
Defect          8          2
No Defect       8        182

In [15]:
confusion_matrix(c3_df.model1, c3_df.actual)

array([[  8,   2],
       [  8, 182]])

 c[0,0]: There are 8   True Positives

c[0:1]: There are 2    False Positives

c[1,0]: There are 8   False Negatives

c[1,1]: There are 182 True Negatives

using recall we get:

In [18]:
print(f'The percentage of positive cases predicted is: {8 / (8 + 8)}')

The percentage of positive cases predicted is: 0.5


In [26]:
c3_df['baseline'] = 'No Defect'

In [27]:
c3_df.head()

actual     model1  model2     model3   baseline
0  No Defect  No Defect  Defect  No Defect  No Defect
1  No Defect  No Defect  Defect     Defect  No Defect
2  No Defect  No Defect  Defect  No Defect  No Defect
3  No Defect     Defect  Defect     Defect  No Defect
4  No Defect  No Defect  Defect  No Defect  No Defect

In [28]:
c3_df.actual == c3_df.model1

0       True
1       True
2       True
3      False
4       True
       ...  
195     True
196     True
197     True
198     True
199     True
Length: 200, dtype: bool

In [29]:
# model1 accuracy
(c3_df.actual == c3_df.model1).mean()

0.95

In [30]:
# baseline accuracy
(c3_df.actual == c3_df.baseline).mean()

0.92

In [33]:
# Model 1
positive = "Defect"

# Subset will be all the times our positive case was correct
subset = c3_df[c3_df.actual == positive]

model_recall = (subset.actual == subset.model1).mean()
baseline_recall = (subset.baseline == subset.actual).mean()

print("Model 1")
print(f"Model recall: {model_recall:.2%}")
print(f"Baseline recall: {baseline_recall:.2%}")

Model 1
Model recall: 50.00%
Baseline recall: 0.00%


In [35]:
# Model 2
positive = "Defect"

# Subset will be all the times our positive case was correct
subset = c3_df[c3_df.actual == positive]

model_recall = (subset.actual == subset.model2).mean()
baseline_recall = (subset.baseline == subset.actual).mean()

print("Model 2")
print(f"Model recall: {model_recall:.2%}")
print(f"Baseline recall: {baseline_recall:.2%}")

Model 2
Model recall: 56.25%
Baseline recall: 0.00%


In [36]:
# Model 3
positive = "Defect"

# Subset will be all the times our positive case was correct
subset = c3_df[c3_df.actual == positive]

model_recall = (subset.actual == subset.model3).mean()
baseline_recall = (subset.baseline == subset.actual).mean()

print("Model 3")
print(f"Model recall: {model_recall:.2%}")
print(f"Baseline recall: {baseline_recall:.2%}")

Model 3
Model recall: 81.25%
Baseline recall: 0.00%


Model 3 should be used.

Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

Doing precision to minimize false positives.

In [40]:
# Model 1
subset = c3_df[c3_df.model1 == positive]
model_precision = (subset.model1 == subset.actual).mean()

subset = c3_df[c3_df.baseline == 'Defect']
baseline_precision = (subset.baseline == subset.actual).mean()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

model precision: 80.00%
baseline precision: nan%


In [41]:
# Model 2
subset = c3_df[c3_df.model2 == positive]
model_precision = (subset.model2 == subset.actual).mean()

subset = c3_df[c3_df.baseline == 'Defect']
baseline_precision = (subset.baseline == subset.actual).mean()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

model precision: 10.00%
baseline precision: nan%


In [42]:
# Model 3
subset = c3_df[c3_df.model3 == positive]
model_precision = (subset.model3 == subset.actual).mean()

subset = c3_df[c3_df.baseline == 'Defect']
baseline_precision = (subset.baseline == subset.actual).mean()

print(f'model precision: {model_precision:.2%}')
print(f'baseline precision: {baseline_precision:.2%}')

model precision: 13.13%
baseline precision: nan%


Clearly Model 1 should be used

4. You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?

In [44]:
pets_df = pd.read_csv('gives_you_paws.csv')
pets_df

actual model1 model2 model3 model4
0       cat    cat    dog    cat    dog
1       dog    dog    cat    cat    dog
2       dog    cat    cat    cat    dog
3       dog    dog    dog    cat    dog
4       cat    cat    cat    dog    dog
...     ...    ...    ...    ...    ...
4995    dog    dog    dog    dog    dog
4996    dog    dog    cat    cat    dog
4997    dog    cat    cat    dog    dog
4998    cat    cat    cat    cat    dog
4999    dog    dog    dog    dog    dog

[5000 rows x 5 columns]

In [45]:
pets_df['baseline'] = 'dog'

In [46]:
pets_df.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [48]:
pets_df.actual.value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [51]:
# Model 1
model_accuracy = (pets_df.model1 == pets_df.actual).mean()
baseline_accuracy = (pets_df.baseline == pets_df.actual).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 80.74%
baseline accuracy: 65.08%


In [52]:
# Model 2
model_accuracy = (pets_df.model2 == pets_df.actual).mean()
baseline_accuracy = (pets_df.baseline == pets_df.actual).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 63.04%
baseline accuracy: 65.08%


In [53]:
# Model 3
model_accuracy = (pets_df.model3 == pets_df.actual).mean()
baseline_accuracy = (pets_df.baseline == pets_df.actual).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 50.96%
baseline accuracy: 65.08%


In [54]:
# Model 4
model_accuracy = (pets_df.model4 == pets_df.actual).mean()
baseline_accuracy = (pets_df.baseline == pets_df.actual).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 74.26%
baseline accuracy: 65.08%


Model 1 should be used for accuracy.  

Models 1 and 4 are better than baseline.

Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?

As above, Model 1 should be used for Phase 1.

In [58]:
# positives are dog

subset = pets_df[pets_df.model1 == 'dog']
model_recall = (subset.model1 == subset.actual).mean()

print(f'model precision: {model_recall:.2%}')


model precision: 89.00%


In [59]:
subset = pets_df[pets_df.model2 == 'dog']
model_recall = (subset.model2 == subset.actual).mean()

print(f'model precision: {model_recall:.2%}')


model precision: 89.32%


In [60]:
subset = pets_df[pets_df.model3 == 'dog']
model_recall = (subset.model3 == subset.actual).mean()

print(f'model precision: {model_recall:.2%}')


model precision: 65.99%


In [61]:
subset = pets_df[pets_df.model4 == 'dog']
model_recall = (subset.model4 == subset.actual).mean()

print(f'model precision: {model_recall:.2%}')


model precision: 73.12%


In [62]:
# positives are cat

subset = pets_df[pets_df.model1 == 'cat']
model_recall = (subset.model1 == subset.actual).mean()

print(f'model precision: {model_recall:.2%}')


model precision: 68.98%


In [63]:
subset = pets_df[pets_df.model2 == 'cat']
model_recall = (subset.model2 == subset.actual).mean()

print(f'model precision: {model_recall:.2%}')

model precision: 48.41%


In [64]:
subset = pets_df[pets_df.model3 == 'cat']
model_recall = (subset.model3 == subset.actual).mean()

print(f'model precision: {model_recall:.2%}')

model precision: 35.83%


In [65]:
subset = pets_df[pets_df.model4 == 'cat']
model_recall = (subset.model4 == subset.actual).mean()

print(f'model precision: {model_recall:.2%}')

model precision: 80.72%


5. Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

sklearn.metrics.accuracy_score


In [67]:
from sklearn.metrics import accuracy_score

In [69]:
y_pred1 = pets_df.model1
y_pred2 = pets_df.model2
y_pred3 = pets_df.model3
y_pred4 = pets_df.model4
y_true = pets_df.actual

In [72]:
accuracy = accuracy_score(y_true,y_pred1)
print(f'Accuracy score for Model1 is {accuracy:.4} ')

Accuracy score for Model1 is 0.8074 


In [73]:
accuracy = accuracy_score(y_true,y_pred2)
print(f'Accuracy score for Model2 is {accuracy:.4} ')

Accuracy score for Model2 is 0.6304 


In [74]:
accuracy = accuracy_score(y_true,y_pred3)
print(f'Accuracy score for Model3 is {accuracy:.4} ')

Accuracy score for Model3 is 0.5096 


In [75]:
accuracy = accuracy_score(y_true,y_pred4)
print(f'Accuracy score for Model4 is {accuracy:.4} ')

Accuracy score for Model4 is 0.7426 


In [77]:
from sklearn.metrics import precision_score

In [78]:
precision = precision_score(y_true, y_pred1, pos_label='dog')
print(f'Precision score for Model1 is {precision:.4} ')

Precision score for Model1 is 0.89 


In [79]:
precision = precision_score(y_true, y_pred2, pos_label='dog')
print(f'Precision score for Model2 is {precision:.4} ')

Precision score for Model2 is 0.8932 


In [80]:
precision = precision_score(y_true, y_pred3, pos_label='dog')
print(f'Precision score for Model3 is {precision:.4} ')

Precision score for Model3 is 0.6599 


In [81]:
precision = precision_score(y_true, y_pred4, pos_label='dog')
print(f'Precision score for Model4 is {precision:.4} ')

Precision score for Model4 is 0.7312 


In [82]:
from sklearn.metrics import recall_score

In [84]:
models = ['model1', 'model2', 'model3', 'model4']

In [85]:
for model in models:
    recall = recall_score(pets_df.actual, pets_df[model], pos_label='dog')
    print(f'Recall score for {model} is {recall:.4}')

Recall score for model1 is 0.8033
Recall score for model2 is 0.4908
Recall score for model3 is 0.5086
Recall score for model4 is 0.9557


In [88]:
from sklearn.metrics import classification_report

In [90]:
for model in models:
    class_report = classification_report(pets_df.actual, pets_df[model], target_names=['dog', 'cat'])
    print(f'CLASSIFICATION REPORT FOR {model.upper()}\n')
    print(class_report)
    print('------------------------------------------')

CLASSIFICATION REPORT FOR MODEL1

              precision    recall  f1-score   support

         dog       0.69      0.82      0.75      1746
         cat       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000

------------------------------------------
CLASSIFICATION REPORT FOR MODEL2

              precision    recall  f1-score   support

         dog       0.48      0.89      0.63      1746
         cat       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000

------------------------------------------
CLASSIFICATION REPORT FOR MODEL3

              precision    recall  f1-score   support

         dog       0.36      0.51      0.42      1746
         cat       0.66      0.51      0.57      